### PR TITLE
[WIP] Add conv2d gradients

### PR DESCRIPTION
**Pull request by nhynes**
This PR enables efficient `conv2d_nchw` gradients on x86. Specifically, it
* adds an x86 schedule for `conv2d_transpose` which nets a 10x boost in performance
* adds the `conv2d_grad_weight_nchw` operator with a custom schedule for x86

Some notes:
* the method of specifying TVM tags should be made 1)  composable and 2) consistent across C++ and Python

WIP until:
- [ ] streamlined functions for all layouts via nchw converters
- [ ] add docs and other pylint necessities

### CODE REVIEW

**Comment by tqchen** [comment]
cc @yzhliu can you do a review?

**Comment by tqchen** [comment]
http://mode-gpu.cs.washington.edu:8080/blue/organizations/jenkins/dmlc%2Ftvm/detail/PR-1075/1/pipeline lint error can be fixed by disable the warning

File: topi/python/topi/x86/conv2d.py [hunk 1]
```diff
@@ -81,81 +81,153 @@ def _declaration_conv(data, kernel, stride, padding, layout, out_dtype):
...
+
+
+def _conv2d_default_schedule(s, conv, data, layout='nchw'):
+    """NCHW conv2d schedule for non imagenet workloads"""
+    data_pad = None
+    if isinstance(data.op, tvm.tensor.ComputeOp) and "pad" in data.op.tag:
+        data_pad = data
+        data = data_pad.op.input_tensors[0]
+    elif isinstance(data.op, tvm.tensor.PlaceholderOp):
+        return
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: why skip scheduling conv ?

File: topi/python/topi/x86/conv2d.py [hunk 2]
```diff
@@ -81,81 +81,153 @@ def _declaration_conv(data, kernel, stride, padding, layout, out_dtype):
...
+    layout: should be a string containing each of the characters {n, c, h, w}
+    """
+    return [tensor.op.axis[i] for oa in 'nchw'
+            for i, ia in enumerate(layout) if oa == ia]
+
+
+def _schedule_pad(schedule, padded_tensor, layout='nchw'):
+    """Fuse and parallelize a padding op."""
+    n, c, h, w = _get_axes_nchw(padded_tensor, layout)
+    pad_fused = schedule[padded_tensor].fuse(n, c)
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: not sure, can we fuse two nonadjacent axis? if run `_get_axes_nchw(padded_tensor, 'nhwc')`, `n` and `c` will be nonadjacent, right?

**Comment by tqchen** [comment]
any updates on this PR?

**Comment by nhynes** [comment]
I'll take a pass at it this week. I'll have to take a non-superficial look at https://discuss.tvmlang.org/t/datalayout-structure/80/2, though.

**Comment by tqchen** [comment]
close for now due to inactive status, please PR new ones when ready :)

**Comment by cbecker** [comment]
Hello. First thanks for creating TVM, it looks very promising to deploy and speed up networks on different architectures.

(disclaimer: I'm very new to TVM, still working on understanding how it works internally)

I'd like to know if there are plans to speed up transposed convolutions, as right now this makes our networks (U-Net-like) run 2 times slower than in pytorch. Alternatively, is there a way to find a workaround to this 

Is this PR also progress towards that https://github.com/dmlc/tvm/issues/2519 ?

**Pull request closed without merging**